src/work/Work.cpp

```diff
@@ -249,25 +266,45 @@ Work::advance()
     }
 
     CLOG(DEBUG, "Work") << "advancing " << getUniqueName();
-    advanceChildren();
-    if (allChildrenSuccessful())
+
+    // If necessary, propagate abort signal before advancing children
+    // This is to prevent scheduling any children to run if they are about
+    // to be in WORK_ABORTING state (such children are scheduled to abort
+    // properly instead)
+    if (anyChildFatalFailure())
     {
-        CLOG(DEBUG, "Work") << "all " << mChildren.size() << " children of "
-                            << getUniqueName() << " successful, scheduling run";
-        scheduleRun();
+        CLOG(DEBUG, "Work")
+            << "some of " << mChildren.size() << " children of "
+            << getUniqueName() << " FATALLY failed, propagating "
+            << "abort";
+        abort(WORK_COMPLETE_FATAL);
     }
-    else if (anyChildFatalFailure())
+    else if (anyChildRaiseFailure())
     {
         CLOG(DEBUG, "Work") << "some of " << mChildren.size() << " children of "
-                            << getUniqueName() << " fatally failed, scheduling "
-                            << "fatal failure";
-        scheduleFatalFailure();
+                            << getUniqueName() << " failed, propagating "
+                            << "abort";
+        abort(WORK_COMPLETE_FAILURE);
```

Comment:
this is basically 'abort all children and then retry', I'm not sure if naming is good here